### PR TITLE
feat(do): infra.spaces_key driver (engine-routing pattern, workflow v0.27.0)

### PR DESCRIPTION
**Pull request by intel352**
## Summary

Lands the `infra.spaces_key` resource driver for the DO plugin under the **engine-routing** pattern (workflow v0.27.0 + ADR 0015/0020). The driver returns `ResourceOutput` with `Sensitive` flags; the engine's `iac/sensitive` package routes flagged fields through the configured `secrets.Provider` per-call. The driver itself is plugin-agnostic — no `secrets.Provider` parameter, no `Set()` calls.

## Tasks

- **Task 11** — failing test for `SpacesKeyDriver.Create` (pre-pivot, replaced this push)
- **Task 12** — `SpacesKeyDriver.Create` + spec→godo translation
- **Task 13** — `SpacesKeyDriver.Read/Update/Diff/Delete` + driver registration
- **Task 14** — failing test for `EnumerateAll` (`infra.spaces_key`)
- **Task 15** — `EnumerateAll` impl with godo pagination
- **Pin bump** — workflow v0.26.0 → v0.27.0 (engine-side sensitive routing)

## Pivot context

This branch was originally on the **pre-pivot** design (driver received a `secrets.Provider` parameter and stored sub-keys directly). Per the user's decision, the design pivoted to engine-routing: drivers return `Outputs` with `Sensitive` flags, the engine handles routing.

The pivot lands as 4 commits forward of `7ad1f02`:
1. `c620f65` — bump workflow to v0.27.0
2. `fd5a1b0` — pivot test fixture
3. `108108f` — implement driver + register
4. `ee28967` — fix stale comment

## Driver contract

| Method | Returns | Sensitive flags |
|---|---|---|
| `Create` | All fields (`name`, `access_key`, `secret_key`, `created_at`, `grants`) | `access_key`, `secret_key` |
| `Read` | All except `secret_key` (DO API only returns it on Create) | `access_key` |
| `Update` | All except `secret_key` | `access_key` |
| `Delete` | — (idempotent, 404 → success) | — |
| `Diff` | Compares mutable fields (name + grants) | — |
| `HealthCheck` | Healthy iff Get succeeds | — |
| `Scale` | `errors.New("not supported")` | — |

`SensitiveKeys()` returns `[access_key, secret_key]` (display masking, independent from per-call routing).
`ProviderIDFormat()` returns `IDFormatFreeform` (DO access keys are not UUIDs).
`SupportsUpsert()` returns `true` (Read can locate by Name when ProviderID is empty).

## Test plan

- [x] `go test ./internal/drivers -run TestSpacesKey` PASSES
- [x] `go test ./...` all pass
- [x] `go build ./...` clean
- [x] `go vet ./...` clean
- [x] `wfctl plugin validate --strict-contracts` — OK
- [x] `iac-codemod refactor-apply -dry-run` — 0 errors

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by intel352** [comment]
# iac-codemod refactor-apply report

Mode:    dry-run
Sites:   1
Errors:  0

## Skipped (// wfctl:skip-iac-codemod)

- internal/provider.go:247 DOProvider.Apply skipped


_Full report (90-day retention) attached as workflow artifact._

**Comment by intel352** [comment]
## Post-pivot rewrite landed (HEAD `ee28967`)

This branch was originally on the **pre-pivot** design (driver took a `secrets.Provider` parameter and called `Set()` itself). It now lands the **post-pivot** engine-routing design per ADR 0015/0020 + workflow v0.27.0 (`iac/sensitive` package, shipped at `ffae1409`).

### Summary of changes since `7ad1f02`

| Commit | Change |
|---|---|
| `c620f65` | Bump workflow to v0.27.0 (engine-side sensitive routing) |
| `fd5a1b0` | Pivot test fixture to engine-routing pattern (drop `secrets.Provider` param + fake) |
| `108108f` | Implement `SpacesKeyDriver` Create/Read/Update/Diff/Delete + register in provider |
| `ee28967` | Update stale Task 14/15 comment in `EnumerateAll` test |

### Driver responsibility under engine-routing

1. `Create` returns `*ResourceOutput` with **all** fields populated including `secret_key`, plus `Sensitive[access_key|secret_key] = true`.
2. The engine's `iac/sensitive.Route` pulls flagged values out, calls `secrets.Provider.Set(ctx, "<resource>_<key>", value)`, replaces them with `secret_ref://...` placeholders before persisting state.
3. The driver is fully **plugin-agnostic** — no `secrets.Provider` parameter, no `Set()` calls.
4. `SensitiveKeys()` (display masking) returns `[access_key, secret_key]`; this is independent of the per-call `Sensitive` map (routing trigger).

### Copilot comments addressed in this push

- **#2** (line 1): switched test to `package drivers_test` (black-box).
- **#4** (line 46): test now validates the request payload (`name` + `grants`) reaching the fake DO server.
- **#5** (line 49): unexpected paths now return HTTP 500 so godo surfaces deterministically.
- **#7** (provider.go:762): `enumerateAllSpacesKeys` now omits `Outputs["grants"]` entirely when empty so the docstring matches reality.
- **#9** (provider_test.go:1146): updated stale "Task 14/15 hasn't landed" comment.
- **#10** (provider.go:704, scope mismatch): the PR title/description still describe the pre-pivot Task 11 + v0.26.0 pin. **Title and description need a rewrite for the new scope** — this branch now ships the full `infra.spaces_key` driver under engine-routing per ADR 0015/0020 + workflow v0.27.0. Will update the PR title/description after CI passes.
- **#1, #3** (compile-fail signals on Task 11 commit): no longer applicable — `NewSpacesKeyDriver` lands in this push and the test compiles + passes.
- **#6** (pagination silent break): already addressed in `7ad1f02` (returns the parse error rather than break-with-partial-result).
- **#8** (`newDOProviderForTest` using `http.DefaultClient`): already addressed in `7ad1f02` (uses `srv.Client()`).

### Verification

- `go build ./...` — clean
- `go test ./...` — all pass (full driver suite + provider tests)
- `go vet ./...` — clean
- `wfctl plugin validate --strict-contracts` — `OK workflow-plugin-digitalocean v0.12.0`
- `iac-codemod refactor-apply -dry-run .` — 0 errors, 1 deliberate skip

File: internal/drivers/spaces_key.go [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+//     `secret_ref://...` placeholders before persisting state. The driver
+//     itself NEVER touches secrets.Provider — it stays platform-agnostic.
+//
+//   - Read returns Outputs WITHOUT secret_key (the DO API doesn't return it
+//     on List/Get; the secret value lives in the secrets.Provider after
+//     Create routed it). Audit/refresh paths reconcile against access_key,
+//     name, created_at, grants — the only fields the API can re-read.
+//
+//   - Diff compares mutable fields (grants). Spaces keys can't be renamed
+//     server-side, so name divergence sets NeedsReplace=true.
```
[the remark this answers is not in the document]
Comment (intel352) [reply internal/drivers/spaces_key.go]: Addressed in 6861817: doc was wrong; godo `SpacesKeysService.Update` accepts both Name and Grants, so the impl is correct (in-place rename). Updated package comment to match — name divergence is NeedsUpdate, not NeedsReplace.

File: internal/drivers/spaces_key.go [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+// spacesKeyOutput renders the standard ResourceOutput for a *godo.SpacesKey.
+// includeSecret controls whether secret_key is populated (true on Create,
+// false on Read/Update where the API doesn't return it).
+func spacesKeyOutput(k *godo.SpacesKey, includeSecret bool) *interfaces.ResourceOutput {
+	outputs := map[string]any{
+		"name":       k.Name,
+		"access_key": k.AccessKey,
+		"created_at": k.CreatedAt,
+		"grants":     grantsToMaps(k.Grants),
+	}
```
[the remark this answers is not in the document]
Comment (intel352) [reply internal/drivers/spaces_key.go]: Addressed in 6861817: `spacesKeyOutput` now omits `Outputs["grants"]` entirely when grants is empty, matching `enumerateAllSpacesKeys`. Both code paths now produce identical state-record shapes.

File: internal/provider.go [hunk 3]
```diff
@@ -679,6 +681,110 @@ func (p *DOProvider) RevokeProviderCredential(ctx context.Context, source string
...
+			if g := grantsToMaps(k.Grants); g != nil {
+				outputs["grants"] = g
+			}
+			all = append(all, &interfaces.ResourceOutput{
+				Name:       k.Name,
+				Type:       "infra.spaces_key",
+				ProviderID: k.AccessKey, // godo identifier — used by Read/Delete
+				Outputs:    outputs,
+				Sensitive:  map[string]bool{"access_key": true},
+				Status:     "running",
```
[the remark this answers is not in the document]
Comment (intel352) [reply internal/provider.go]: Addressed in 6861817: `enumerateAllSpacesKeys` now sets Status="active" to match `SpacesKeyDriver.Create` and the sister bucket driver (`internal/drivers/spaces.go`). Added a comment explaining the canonical choice (Spaces keys have no lifecycle status in the godo API).

File: internal/provider_test.go [hunk 4]
```diff
@@ -1096,3 +1099,116 @@ func TestDOProvider_Apply_UpsertAllDrivers(t *testing.T) {
...
+				return
+			}
+			_ = json.NewEncoder(w).Encode(map[string]any{
+				"keys": []any{
+					map[string]any{"name": "key-c", "access_key": "AK_C", "created_at": "2026-05-03T00:00:00Z"},
+				},
+			})
+			return
+		}
+		t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
```
[the remark this answers is not in the document]
Comment (intel352) [reply internal/provider_test.go:1177]: Addressed in 6861817: handler now calls `http.Error(w, "unexpected path", http.StatusInternalServerError)` after `t.Errorf` so godo surfaces the failure as an HTTP error rather than potentially decoding an empty 200 body. Mirrors the same pattern in `internal/drivers/spaces_key_test.go`.

File: internal/drivers/spaces_key.go [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+	key, _, err := d.client.Get(ctx, ref.ProviderID)
+	if err != nil {
+		return nil, fmt.Errorf("spaces_key read %q: %w", ref.Name, WrapGodoError(err))
+	}
+	if key == nil {
+		return nil, fmt.Errorf("spaces_key read %q: godo returned nil key", ref.Name)
+	}
+	return spacesKeyOutput(key, /*includeSecret*/ false), nil
+}
```
[the remark this answers is not in the document]
Comment (intel352) [reply internal/drivers/spaces_key.go:126]: Addressed in 6861817: tightened the package doc to state explicitly that BOTH Read and Update return Outputs without secret_key (Update was previously implicit), and that Delete is idempotent on 404. Verified against impl: Read/Update both call `spacesKeyOutput(key, /*includeSecret*/ false)`; Delete checks `resp != nil && resp.StatusCode == 404` and returns nil.

[the remark this answers is not in the document]
Comment (intel352) [reply internal/drivers/spaces_key.go:126]: Addressed in 4e3f0ed: 15 new tests added covering Read (4 paths: happy + 404 + findByName + findByName-not-found), Update (rename-in-place + grants-only), Delete (happy + 404-idempotent + 500-propagates), Diff (nil-current + no-changes + name-change + grants-change + structpb-roundtrip-resilience), and HealthCheck. All assert via `errors.Is` against `interfaces.Err*` sentinels where applicable, and pin the secret_key-absent-on-Read/Update contract that the package doc calls out. The structpb-roundtrip test specifically guards the gRPC plugin boundary case where `Outputs["grants"]` decodes as `[]any{map[string]any}` instead of `[]map[string]any` — `normalizeGrantsForDiff` handles that and Diff no longer false-positives a no-op refresh as drift.

**Comment by intel352** [comment]
## Test-coverage gap closed

**Commit:** [4e3f0ed](https://github.com/GoCodeAlone/workflow-plugin-digitalocean/commit/4e3f0ed2dfa645be6e1fa3e7586af2a94015a5ec) — `test(spaces_key): add Read/Update/Delete/Diff/HealthCheck lifecycle coverage`

**15 new tests** in `internal/drivers/spaces_key_test.go` (+655 lines), all stdlib `testing` + `httptest` (no testify):

| Method | Tests |
|---|---|
| Read | HappyPath, NotFound (errors.Is ErrResourceNotFound), FindByName (paginated fallback), FindByName_NotFound |
| Update | RenameInPlace, GrantsChange |
| Delete | HappyPath (204), Idempotent_404, OtherError_Propagates (500 → ErrTransient) |
| Diff | NilCurrent_NeedsUpdate, NoChanges, NameChange (NeedsUpdate not NeedsReplace per ADR 0015), GrantsChange, **StructpbRoundtripResilience** |
| HealthCheck | HealthCheck (existence == healthy) |

**Critical correctness checks pinned:**
- `secret_key` MUST NOT appear in Outputs on Read or Update (DO API never re-emits it after Create — engine-side routing already handled the secret)
- `Sensitive[secret_key]` unset on Read/Update; only `access_key` flagged
- Delete idempotency on 404 (matches `RevokeProviderCredential`)
- Delete propagates non-404 errors via `ErrTransient` (no silent swallow)
- **Diff resilient to structpb gRPC-boundary roundtrip:** `Outputs["grants"]` as `[]any{map[string]any}` does NOT false-positive as drift (per workspace memory `feedback_workflow_plugin_structpb_boundary`)

**Local verification (HEAD 4e3f0ed):**
```
GOWORK=off go test ./internal/drivers -count=1  → ok
GOWORK=off go test ./internal -count=1          → ok
GOWORK=off go build ./...                       → ok
```

No driver-impl changes were necessary — all tests pass against the existing implementation, confirming the v0.27.0 engine-routing contract is correctly implemented.
